R-package/tests/testthat/test_basic.R

```diff
@@ -188,6 +188,60 @@ test_that("lightgbm() rejects negative or 0 value passed to nrounds", {
   }
 })
 
+test_that("lightgbm() accepts nrounds as either a top-level argument or parameter", {
+  nrounds <- 15L
+
+  set.seed(708L)
+  top_level_bst <- lightgbm(
+    data = train$data
+    , label = train$label
+    , nrounds = nrounds
+    , params = list(
+      objective = "regression"
+      , metric = "l2"
+      , num_leaves = 5L
+    )
+    , save_name = tempfile(fileext = ".model")
+  )
+
+  set.seed(708L)
+  param_bst <- lightgbm(
+    data = train$data
+    , label = train$label
+    , params = list(
+      objective = "regression"
+      , metric = "l2"
+      , num_leaves = 5L
+      , nrounds = nrounds
+    )
+    , save_name = tempfile(fileext = ".model")
+  )
+
+  set.seed(708L)
+  both_customized <- lightgbm(
+    data = train$data
+    , label = train$label
+    , nrounds = 20L
+    , params = list(
+      objective = "regression"
+      , metric = "l2"
+      , num_leaves = 5L
+      , nrounds = nrounds
+    )
+    , save_name = tempfile(fileext = ".model")
+  )
+
+  expect_equal(param_bst$current_iter(), top_level_bst$current_iter())
+  expect_equal(param_bst$best_score
+               , top_level_bst$best_score
+               , tolerance = TOLERANCE)
```

Comment:
I don't think `best_score` is the right thing to use here. For any LightGBM training where validation data isn't provided, this will always be `NA`.

```r
library(lightgbm)

data(agaricus.train, package = "lightgbm")
train <- agaricus.train

nrounds <- 15L

top_level_bst <- lightgbm(
    data = train$data
    , label = train$label
    , nrounds = nrounds
    , params = list(
      objective = "regression"
      , metric = "l2"
      , num_leaves = 5L
    )
    , save_name = tempfile(fileext = ".model")
)

top_level_bst$best_score
# [1] NA
```

To test "the model produced is identical", to be more confident that the value of `num_iterations` was passed through, you can use `$eval_train()`. This creates predictions on the training data, using the full model, and then provides evaluation metrics based on those predictions.

```r
top_level_l2 <- top_level_bst$eval_train()[[1L]][["value"]]
params_l2    <- params_bst$eval_train()[[1L]][["value"]]

# check type just to be sure the subsetting didn't return a NULL
expect_true(is.numeric(top_level_2))
expect_true(is.numeric(params_l2))

# check that model produces identical performance
expect_equal(top_level_l2, params_l2)
```

Could you make this change?